src/common/util/richVarinfo.ml

```diff
@@ -17,10 +10,78 @@ sig
   val bindings: unit -> (t * varinfo) list
 end
 
+module VarinfoDescription = struct
+  (**This type is equal to `varinfo`, but the fields are not mutable and they are optional.
+     Only the name is not optional. *)
+  type t = {
+    vname_: string;
+    vtype_: typ option;
+    vattr_: attributes option;
+    vstorage_: storage option;
+    vglob_: bool option;
+    vinline_: bool option;
+    vdecl_: location option;
+    vinit_: initinfo option;
+    vaddrof_: bool option;
+    vreferenced_: bool option;
+  }
+
+  let from_varinfo (v: varinfo) =
+    {vname_=v.vname;
+     vtype_=Some v.vtype;
+     vattr_=Some v.vattr;
+     vstorage_=Some v.vstorage;
+     vglob_=Some v.vglob;
+     vinline_=Some v.vinline;
+     vdecl_=Some v.vdecl;
+     vinit_=Some v.vinit;
+     vaddrof_=Some v.vaddrof;
+     vreferenced_=Some v.vreferenced}
+
+  let empty name =
+    {vname_=name;
+     vtype_=None;
+     vattr_=None;
+     vstorage_=None;
+     vglob_=None;
+     vinline_=None;
+     vdecl_=None;
+     vinit_=None;
+     vaddrof_=None;
+     vreferenced_=None}
+
+  let update_varinfo (v: varinfo) (update: t) =
+    let open Batteries in
+    {vname=update.vname_;
+     vtype=Option.default v.vtype update.vtype_;
+     vattr=Option.default v.vattr update.vattr_;
+     vstorage=Option.default v.vstorage update.vstorage_;
+     vglob=Option.default v.vglob update.vglob_;
+     vinline=Option.default v.vinline update.vinline_;
+     vdecl=Option.default v.vdecl update.vdecl_;
+     vinit=Option.default v.vinit update.vinit_;
+     vid=v.vid;
+     vaddrof=Option.default v.vaddrof update.vaddrof_;
+     vreferenced=Option.default v.vreferenced update.vreferenced_;
+     vdescr=v.vdescr;
+     vdescrpure=v.vdescrpure;
+     vhasdeclinstruction=v.vhasdeclinstruction}
+end
```

Comment:
This seems like a massive amount of boilerplate for what exactly? The fields of `varinfo` _are_ mutable, so the type could just be changed in-place with `v.vtype <- t` without introducing copies of the `varinfo` with the same `vid` but different data.

When `varinfo`s are correctly compared with `CilType.Varinfo`, then that only depends on `vid` anyway, so such mutations shouldn't break any data structures they are in either.